asm/opcode.go

```diff
@@ -125,11 +129,20 @@ func (op OpCode) Endianness() Endianness {
 }
 
 // JumpOp returns the JumpOp.
+// Returns InvalidJumpOp if it doesn't encode a jump.
 func (op OpCode) JumpOp() JumpOp {
-	if op.Class().encoding() != jumpOrALU {
+	if op.Class() != JumpClass && op.Class() != Jump32Class {
 		return InvalidJumpOp
 	}
-	return JumpOp(op & jumpMask)
+
+	jumpOp := JumpOp(op & jumpMask)
+
+	// Some JumpOps are only supported by JumpClass, not Jump32Class.
+	if op.Class() == Jump32Class && (jumpOp == Exit || jumpOp == Call || jumpOp == Ja) {
```

Comment:
I don't think we should do this here. There is an asymmetry in using `asm`:
* When using it to assemble programs we shouldn't allow constructing invalid programs, so `SetJumpOp` should return Invalid.
* When using it to introspect programs we should allow decoding instructions as much as possible. `JumpOp` should do the minimum checks necessary.

Reply:
The kernel docs specifically call them out as BPF_JUMP only: https://www.kernel.org/doc/Documentation/networking/filter.txt

```
  BPF_JA    0x00  /* BPF_JMP only */
  BPF_CALL  0x80  /* eBPF BPF_JMP only: function call */
  BPF_EXIT  0x90  /* eBPF BPF_JMP only: function return */
```

Right now the kernel rejects them. If it does accept them in the future, I'm half worried they'll be abused for some different purpose - it feels like labeling them as Invalid / Unknown would help a user notice something is up. But I'm happy to change it if you feel differently.

Reply:
Ok, there are a bunch of places in the library which use JumpOp to to function rewrity things, so returning invalidJumpOp here is the cautious option.